### DISCUSSION TITLE
How to activate Cache for  file attributes

### DISCUSSION THREAD

**Discussion by christophgil** [libfuse/libfuse#832] [Q&A]
Dear FUSE developers,

Many thanks for maintaining this excellent library.

Even though I am not a professional C programmer, I found it easy and great fun to get my file system working with
the documentation and the examples in particular passthrough.c.

Please let me ask some questions regarding performance and design.

## Questions on caching struct stat

I need to reduce the number of stat calls. My understanding is that the struct stat data which is passed by   

       filler(buf, de->d_name, &st, 0, fill_dir_plus)

can go into the cache of libfuse and reused within a specified time.

For testing I put large timouts in  xmp_init(struct fuse_conn_info *conn,struct fuse_config *cfg):

      cfg->entry_timeout=cfg->attr_timeout=200;

In xmp_readdir(const char *path, void *buf, fuse_fill_dir_t filler,off_t offset, struct fuse_file_info *fi,enum fuse_readdir_flags flags)
It set the following flags:

      fi->keep_cache=1;  fi->cache_readdir=1;

With these settings, I was expecting that xmp_getattr() is less often called. Unfortunately, there is no reduction.
What am I doing wrong?

## Question on direct io

Can I put   

      fi->direct_io=1

  into 

      xmp_read(const char *path, char *buf, size_t size, off_t offset,struct fuse_file_info *fi)?

Or does it belong into  xmp_open()?

## Some questions on design:

- What is the use of xmp_seek() since  xmp_read() has a parameter off_t offset?

- What is the use of xmp_access() since AFAIK this could be implemented with the xmp_getattr() data.

- Is it correct to use the last-modified-attribute of files (st_mtime) for checking whether cache
  entries are up-to-date?  Time has some granularity and in theory two changes to a file could
  happen within the same st_mtime value. Even very unlikely, it seems to me that it is possible that
  a cache implementation may fail to recognize that cached data is not valid any more.  
  This could happen on a FS without nanoseconds precision if two changes happen in less than a millisec time.
  AFAIK a correct implementation would use a modification count which would increment whenever there is a
  change to a file. I would have expected such modification counter in struct stat, but there is no
  such field.  Is this a design error of Linux/UNIX file API or am I missing something?

Best
Christoph

**Comment by Nikratio** [comment]
Thanks for your interest in libfuse! As far as I can tell, this is a question about how to use fuse rather than a bug report. Could you please send your question to the FUSE mailing list (https://lists.sourceforge.net/lists/listinfo/fuse-devel), instead of using the issue tracker? I'd like to reserve use of the latter to actually track issues and not use it for discussion.

Thanks!

**Comment by trapexit** [comment]
@Nikratio you have discussions enabled here so might be better to just move these tickets to discussions. You can find the option in the bottom right of the page.

**Comment by GiorgosXou** [comment]
I was wondering *(roughly)* the same thing. Specifically, how could I store\cache the inodes in `readdir` such that I could later use them via `fi` in `getattr` and so on...? *(just like filenames are stored via `filler`)* 

*<sup>context: I'm trying to map a set of inodes to a vfs</sup>*
